paramiko/client.py

```diff
@@ -605,17 +609,29 @@ def _auth(self, username, password, pkey, key_filenames, allow_agent,
 
         if password is not None:
             try:
-                self._transport.auth_password(username, password)
-                return
+                self._log(DEBUG, 'Trying password auth')
+                allowed_types = set(self._transport.auth_password(username, password))
+                two_factor = (allowed_types & two_factor_types)
+                if not two_factor:
+                    return
             except SSHException as e:
                 saved_exception = e
         elif two_factor:
             try:
-                self._transport.auth_interactive_dumb(username)
+                self._transport.auth_interactive_dumb(username, handler=interactive_handler)
```

Comment:
It looks like this `elif two_factor:` block is redundant with  the `if two_factor:` block below. I think you can remove this one.

Further, with this block as it is, if password is None and `two_factor` is triggered, it looks like `auth_interactive_dumb()` will be called twice.